// @aws-blocks/just-in-time-registration/lambdas/registration-lambda/lib/greengrass/greengrass-provisioner.js
const _                = require('lodash');
const AWS              = require('aws-sdk');
const loggerDefinition = require('./logger-definition');

// Creating a new `greengrass` client.
const greengrass = new AWS.Greengrass();

/**
 * Creates a new logger definition to be used
 * by new Greengrass Groups.
 */
const createLoggerDefinition = () => {
  return (new Promise((resolve, reject) => {
    greengrass.createLoggerDefinition({
      InitialVersion: {
        Loggers: loggerDefinition
      }
    }, (err, data) => (err ? reject(err) : resolve(data.LatestVersionArn)));
  }));
};

/**
 * Creates a new Greengrass core definition.
 * @param {*} thingArn the ARN of the thing to associate with the
 * core definition.
 * @param {*} certificateArn the ARN of the certificate associated
 * with the thing.
 */
const createCoreDefinition = (thing, certificateArn) => {
  return (new Promise((resolve, reject) => {
    greengrass.createCoreDefinition({
      InitialVersion: {
        Cores: [{
          CertificateArn: certificateArn,
          Id: thing.thingName,
          ThingArn: thing.thingArn
        }]
      },
      Name: thing.thingArn
    }, (err, data) => (err ? reject(err) : resolve(data.LatestVersionArn)));
  }));
};

/**
 * Creates a new Greengrass group.
 * @param {*} name the thing name associated with the
 * Greengrass device.
 * @param {*} core the ARN of the core to associate.
 * @param {*} loggers the ARN of the loggers to associate.
 */
const createGroup = (name, coreArn, loggerArn) => {
  return (new Promise((resolve, reject) => {
    greengrass.createGroup({
      InitialVersion: {
        CoreDefinitionVersionArn: coreArn,
        LoggerDefinitionVersionArn: loggerArn
      },
      Name: name
    }, (err, data) => (err ? reject(err) : resolve(data)));
  }));
};

/**
 * Associates a role with the current account.
 * @param {*} roleArn the ARN of the role to associate as a service
 * role to the current account.
 */
const associateServiceRoleToAccount = (roleArn) => {
  return (new Promise((resolve, reject) => {
    greengrass.associateServiceRoleToAccount({
      RoleArn: roleArn
    }, (err, data) => err ? reject(err) : resolve(data));
  }));
};

/**
 * Provisions a new Greengrass Core and Group if the
 * given device is a `greengrass` device.
 */
module.exports.provision = (thing, certificate) => {
  if (certificate.attributes.subject.T !== 'greengrass') {
    // The device is not a `greengrass` device.
    return (Promise.resolve());
  }
  // The name of the Greengrass group.
  const group = _.template(process.env.GreengrassGroupName)({ certificate });
  // The resource description of the created definitions.
  const results = {};
  // Starting by verifying whether a Greengrass service role exists on the current account.
  return (createCoreDefinition(thing, certificate.certificateDescription.certificateArn)
    // Storing the core ARN.
    .then((coreArn) => results.coreArn = coreArn)
    // Creating a new Greengrass Logger definition.
    .then(() => createLoggerDefinition())
    // Storing the logger ARN.
    .then((loggerArn) => results.loggerArn = loggerArn)
    // Associating the service role with the account.
    .then(() => associateServiceRoleToAccount(process.env.GreengrassServiceRole))
    // Creating a new Greengrass Group definition.
    .then(() => createGroup(group, results.coreArn, results.loggerArn))
    // Storing the group ARN.
    .then((group) => {
      // Saving the group definition.
      results.group = group;
      return (results);
    })
  );
};
